### ISSUE TITLE
tf.keras.BatchNormalization ignores Mask when calculating Mean/Variance

### ISSUE THREAD

**Issue by LukeBolly** [tensorflow/tensorflow#30161]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 1.13.1
- Python version: 3.6.7
- CUDA/cuDNN version: 10
- GPU model and memory: GTX 960M

**Describe the current behavior**
tf.keras.layers.BatchNormalization layers do not take the Mask into account when calculating mean/variance. This results in a Mean biased towards 0 and an inflated Variance when passing in padded samples. This is made worse by the fact that the BatchNormalization layer claims to support masking, yet all it does is pass it through to the next layer. This is an issue in TF 1.13, 1.14 and 2.0b where the implementation is imported straight from Keras.

The result from a toy example:
```
padded mean:		1.5
unpadded mean:		3.0

padded variance:	3.25
unpadded variance:	2.0
```

**Describe the expected behavior**
The Boolean Mask should be used when calculating the Mean and Variance, to ignore any 0's as a result of padded data.

**Code to reproduce the issue**
This example compares the behavior of BatchNorm against a padded and un-padded 1D signal. The resulting Means and Variances come out very different.
```
import tensorflow as tf
import numpy as np

# Mask value of -10
padded_data = np.expand_dims(np.array([[1,2,3,4,5,-10,-10,-10,-10,-10] for _ in range(100)], dtype='float32'), 2)
data = np.expand_dims(np.array([[1,2,3,4,5] for _ in range(100)], dtype='float32'), 2)

def build_padding_model():
    input = tf.keras.layers.Input((10, 1))
    masked = tf.keras.layers.Masking(-10)(input)
    normed = tf.keras.layers.BatchNormalization(momentum=0.01)(masked)
    model = tf.keras.models.Model(input, normed)
    model.compile("adam", loss="mse")
    return model

def build_model():
    input = tf.keras.layers.Input((5, 1))
    normed = tf.keras.layers.BatchNormalization(momentum=0.01)(input)
    model = tf.keras.models.Model(inputs=input, outputs=normed)
    model.compile("adam", loss="mse")
    return model

if __name__ =='__main__':
    batch_size = 100
    signal_length = 20

    pad = build_padding_model()
    pad.fit(x=padded_data, y=padded_data, batch_size=10, epochs=5)

    nopad = build_model()
    nopad.fit(x=data, y=data, batch_size=10, epochs=5)

    weights1 = pad.layers[2].get_weights()
    weights2 = nopad.layers[1].get_weights()
    print('\npadded mean:\t\t' + str(weights1[2][0]) + '\nunpadded mean:\t\t' + str(weights2[2][0]) + '\n')
    print('padded variance:\t' + str(weights1[3][0]) + '\nunpadded variance:\t' + str(weights2[3][0]) + '\n')
```

**Comment by LukeBolly** [comment]
For what it's worth, I have modified a version of the 1.13.1 BatchNormalization layer to take a mask in `call` which appears to be working. The change wasn't very big, though I haven't done anything to support renorm/fused/virtual/adjustment configurations. Would this be something worth creating a PR for?

**Comment by ravikyram** [comment]
I have tried on colab with TF  1.13.1, 1.14 and 2.0b and was able to reproduce the issue.Thanks!

**Comment by karmel** [comment]
The trick here is making sure we do this in a performant way, as the op being called underneath does not support masking. We could consider going a slower route, filtering out the masked items, in the cases where a mask exists. If you wanted to send a PR, we could take a look and see how bad the performance hit is/what the trade-offs are.

**Comment by georgesterpu** [comment]
Hi @karmel 
Since there hasn't been any progress on the masked batch norm PR lately, I would like to ask about a cheaper workaround.

Instead of zero padding the shorter sequences, one option is to sample from the existing sequences, e.g. by mirroring/replicating samples from the beginning/end. Could this be done conveniently in the `padded_batch` method of `tf.data.Dataset` with the `padding_values` parameter ?

**Comment by karmel** [comment]
@georgesterpu -- a lot has changed since this issue was first posted. Is this still an issue in TF v2? Can you try with tf-nightly or the most recent release of tensorflow?

**Comment by georgesterpu** [comment]
@karmel thanks for the follow up.
Yes, this feature is not yet supported in tf-nightly, and I guess it is still an issue for anyone using the batch normalisation op on zero-padded sequences.

**Comment by hoangcuong2011** [comment]
@LukeBolly Thx for bring this up. I feel this an important feature to be added.
BTW I have a question: I think this also a problem with LayerNormalization, where it does not use mask when calculating Mean/Variance. Please verify or correct me, if possible. @LukeBolly @georgesterpu Thx!

**Comment by hoangcuong2011** [comment]
OK after some research, I found LayerNormalization USES mask when computing these things. I modified the code of @LukeBolly as below to see this. Thx.

```
import tensorflow as tf
import numpy as np

# Mask value of -10
padded_data = np.expand_dims(np.array([[1,2,3,4,5,-10,-10,-10,-10,-10] for _ in range(1000)], dtype='float32'), 2)
data = np.expand_dims(np.array([[1,2,3,4,5] for _ in range(1000)], dtype='float32'), 2)

def build_padding_model():
    input = tf.keras.layers.Input((10, 1))
    masked = tf.keras.layers.Masking(-10)(input)
    normed = tf.keras.layers.LayerNormalization()(masked)
    model = tf.keras.models.Model(input, normed)
    model.compile("adam", loss="mse")
    return model

def build_model():
    input = tf.keras.layers.Input((5, 1))
    normed = tf.keras.layers.LayerNormalization()(input)
    model = tf.keras.models.Model(inputs=input, outputs=normed)
    model.compile("adam", loss="mse")
    return model

if __name__ =='__main__':
    batch_size = 100
    signal_length = 20

    pad = build_padding_model()
    pad.fit(x=padded_data, y=padded_data, batch_size=10, epochs=5)

    nopad = build_model()
    nopad.fit(x=data, y=data, batch_size=10, epochs=5)

    weights1 = pad.get_weights()
    weights2 = nopad.get_weights()

    print("\npadded mean" + str(weights1) + '\n')
    print("\nopad mean" + str(weights2) + '\n')
```

**Comment by mathpluscode** [comment]
> OK after some research, I found LayerNormalization USES mask when computing these things. I modified the code of @LukeBolly as below to see this. Thx.
> 
> ```
> import tensorflow as tf
> import numpy as np
> 
> # Mask value of -10
> padded_data = np.expand_dims(np.array([[1,2,3,4,5,-10,-10,-10,-10,-10] for _ in range(1000)], dtype='float32'), 2)
> data = np.expand_dims(np.array([[1,2,3,4,5] for _ in range(1000)], dtype='float32'), 2)
> 
> def build_padding_model():
>     input = tf.keras.layers.Input((10, 1))
>     masked = tf.keras.layers.Masking(-10)(input)
>     normed = tf.keras.layers.LayerNormalization()(masked)
>     model = tf.keras.models.Model(input, normed)
>     model.compile("adam", loss="mse")
>     return model
> 
> def build_model():
>     input = tf.keras.layers.Input((5, 1))
>     normed = tf.keras.layers.LayerNormalization()(input)
>     model = tf.keras.models.Model(inputs=input, outputs=normed)
>     model.compile("adam", loss="mse")
>     return model
> 
> if __name__ =='__main__':
>     batch_size = 100
>     signal_length = 20
> 
>     pad = build_padding_model()
>     pad.fit(x=padded_data, y=padded_data, batch_size=10, epochs=5)
> 
>     nopad = build_model()
>     nopad.fit(x=data, y=data, batch_size=10, epochs=5)
> 
>     weights1 = pad.get_weights()
>     weights2 = nopad.get_weights()
> 
>     print("\npadded mean" + str(weights1) + '\n')
>     print("\nopad mean" + str(weights2) + '\n')
> ```

I tried this code in colab with tf2.3.0 and the padded mean and nopad mean are exactly the same. However, if we change it to batch norm, the results are different.

Moreover, for layer norm we could not pass a mask explicitly when calling it. This inconsistent behaviour is really confusing.

**Comment by hoangcuong2011** [comment]
@mathpluscode: Hi,

"However, if we change it to batch norm, the results are different." -> Batch norm in TF implementation does not take masking into account I think.
"Moreover, for layer norm we could not pass a mask explicitly when calling it" -> We can do that actually - just created a custom layer and pass a mask you want thorough that layer. I did it before.

**Comment by mathpluscode** [comment]
> @mathpluscode: Hi,
> 
> "However, if we change it to batch norm, the results are different." -> Batch norm in TF implementation does not take masking into account I think.
> "Moreover, for layer norm we could not pass a mask explicitly when calling it" -> We can do that actually - just created a custom layer and pass a mask you want thorough that layer. I did it before.

Hi @hoangcuong2011, thanks for your reply!

Are you saying that we can write a custom layer to wrap the `tf.keras.layers.LayerNormalization` and use `tf.keras.layers.Masking` to use the given mask? Ideally, I'd like to call like `layer_norm(inputs=x, mask=mask)` to use layer norm.

I'm not sure if I understand your meaning accurately. If it's possible, do you mind to share a piece of the code ;) Super thx!

**Comment by hoangcuong2011** [comment]
"Are you saying that we can write a custom layer to wrap the tf.keras.layers.LayerNormalization and use tf.keras.layers.Masking to use the given mask?" -> Yes this is what I meant. I previously wrote a custom layer that has tf.keras.layers.LayerNormalization inside. If you read about custom layer you might notice that we can pass whatever mask we want inside the custom layer.

About the code: Sorry I did this for my previous company so I cannot share the code. But it is not that hard believe me. Plus you could learn something from doing that.

**Comment by Saduf2019** [comment]
@LukeBolly 
Is this still an issue,Could you check on later versions of 2.x  as 1.x is not supported actively

**Comment by vlawhern** [comment]
I know this is a pretty old thread but just wanted to add some more info for this for those who may still be interested...

I tested the PR from @LukeBolly that implements the masking in BatchNormalization (https://github.com/tensorflow/tensorflow/pull/30298) in TF 2.4 and it works when the input is 1D; however when it's a 2D input you have to pass in ```fused=False``` when defining the ```BatchNormalization``` layer. For example...

```python
import tensorflow as tf
import numpy as np

# Mask value of -10
padded_data = np.expand_dims(np.array([[1,2,3,4,5,-10,-10,-10,-10,-10,6,7,8,9,10,-10,-10,-10,-10,-10] for _ in range(1000)], dtype='float32'), 2)
padded_data = padded_data.reshape(1000, 2, 10, 1)
data = np.expand_dims(np.array([[1,2,3,4,5,6,7,8,9,10] for _ in range(1000)], dtype='float32'), 2)
data = data.reshape(1000, 2, 5, 1)


def build_padding_model():
    input = tf.keras.layers.Input((2, 10, 1))
    masked = tf.keras.layers.Masking(-10)(input)
    normed = tf.keras.layers.BatchNormalization(fused=False)(masked)
    model = tf.keras.models.Model(input, normed)
    model.compile("adam", loss="mse")
    return model

def build_model():
    input = tf.keras.layers.Input((2, 5, 1))
    normed = tf.keras.layers.BatchNormalization()(input)
    model = tf.keras.models.Model(inputs=input, outputs=normed)
    model.compile("adam", loss="mse")
    return model

if __name__ =='__main__':
    batch_size = 100
    signal_length = 20
    
    pad = build_padding_model()
    pad.fit(x=padded_data, y=padded_data, batch_size=10, epochs=5)
    
    nopad = build_model()
    nopad.fit(x=data, y=data, batch_size=10, epochs=5)
    
    weights1 = pad.get_weights()
    weights2 = nopad.get_weights()
    
    print("\npadded mean" + str(weights1) + '\n')
    print("\nopad mean" + str(weights2) + '\n')

padded mean[array([1.4691201], dtype=float32), array([0.48987818], dtype=float32), array([5.4638605], dtype=float32), array([8.202366], dtype=float32)]


opad mean[array([1.46912], dtype=float32), array([0.48987824], dtype=float32), array([5.4638658], dtype=float32), array([8.285133], dtype=float32)]
```
As you can see it's "mostly" right (only minor difference in the last parameter).

Compare this to without the ```fused=False``` flag:

```python
import tensorflow as tf
import numpy as np

# Mask value of -10
padded_data = np.expand_dims(np.array([[1,2,3,4,5,-10,-10,-10,-10,-10,6,7,8,9,10,-10,-10,-10,-10,-10] for _ in range(1000)], dtype='float32'), 2)
padded_data = padded_data.reshape(1000, 2, 10, 1)
data = np.expand_dims(np.array([[1,2,3,4,5,6,7,8,9,10] for _ in range(1000)], dtype='float32'), 2)
data = data.reshape(1000, 2, 5, 1)


def build_padding_model():
    input = tf.keras.layers.Input((2, 10, 1))
    masked = tf.keras.layers.Masking(-10)(input)
    normed = tf.keras.layers.BatchNormalization()(masked)
    model = tf.keras.models.Model(input, normed)
    model.compile("adam", loss="mse")
    return model

def build_model():
    input = tf.keras.layers.Input((2, 5, 1))
    normed = tf.keras.layers.BatchNormalization()(input)
    model = tf.keras.models.Model(inputs=input, outputs=normed)
    model.compile("adam", loss="mse")
    return model

if __name__ =='__main__':
    batch_size = 100
    signal_length = 20
    
    pad = build_padding_model()
    pad.fit(x=padded_data, y=padded_data, batch_size=10, epochs=5)
    
    nopad = build_model()
    nopad.fit(x=data, y=data, batch_size=10, epochs=5)
    
    weights1 = pad.get_weights()
    weights2 = nopad.get_weights()
    
    print("\npadded mean" + str(weights1) + '\n')
    print("\nopad mean" + str(weights2) + '\n')

padded mean[array([1.4775826], dtype=float32), array([0.47813976], dtype=float32), array([2.7319329], dtype=float32), array([11.675619], dtype=float32)]


opad mean[array([1.46912], dtype=float32), array([0.48987824], dtype=float32), array([5.4638658], dtype=float32), array([8.285133], dtype=float32)]
```

Note that this comes with some drawbacks, namely significantly longer compute time (~30-40% longer per batch with ```fused=False``` in my own analyses.. YMMV). I also haven't tried making the same changes with ```SyncBatchNormalization``` so that you can do distributed GPU training with ```tf.distribute.MirroredStrategy``` for example. But this PR gets me the functionality I need for my projects.

Thanks again to @LukeBolly

**Comment by albertz** [comment]
@ymodak This is still an issue. Can someone reopen this?
